src/xpcc/driver/position/vl53l0_impl.hpp

```diff
@@ -0,0 +1,641 @@
+
+	VL53L0_RF_CALL(read(Register::MSRC__CONFIG_TIMEOUT_MACROP, sequenceInfo.msrcDssTccTimeout));
+	sequenceInfo.msrcDssTccTimeout += 1;
+
+	VL53L0_RF_CALL(read(Register::PRE_RANGE__CONFIG_TIMEOUT_MACROP_HI, &i2cBuffer[1], 2));
+	sequenceInfo.preRangeTimeout = decodeTimeout((i2cBuffer[1] << 8) + i2cBuffer[2]);
+
+	VL53L0_RF_CALL(read(Register::FINAL_RANGE__CONFIG_TIMEOUT_MACROP_HI, &i2cBuffer[1], 2));	
+	sequenceInfo.finalRangeTimeout = decodeTimeout((i2cBuffer[1] << 8) + i2cBuffer[2]);
+
+	RF_END_RETURN(true);
+}
+
+template < typename I2cMaster >
+bool
+xpcc::Vl53l0<I2cMaster>::calculateFinalRangeTimeout(uint32_t measurementTime, uint16_t& registerValue)
+{
+	if(measurementTime < TimeOverhead::Start + TimeOverhead::End) {
+		return false;
+	}
+
+	// available time for "final range" step in microseconds
+	uint32_t finalRangeTime = measurementTime - TimeOverhead::Start - TimeOverhead::End;
+	const uint16_t msrcDssTccTimeout = timeoutToMicroseconds(sequenceInfo.msrcDssTccTimeout,
+															 sequenceInfo.vcselPeriodPreRange);
+
+	if(sequenceInfo.enabledSteps & MeasurementSequenceStep::TCC)
+	{
+		const uint16_t stepTime = TimeOverhead::TCC + msrcDssTccTimeout;
+		if(finalRangeTime <= stepTime) {
+			return false;
+		}
+		finalRangeTime -= stepTime;
+	}
+
+	// "dynamic spad selection" is enabled
+	if(sequenceInfo.enabledSteps & MeasurementSequenceStep::DSS)
+	{
+		const uint16_t stepTime = 2 * (TimeOverhead::DSS + msrcDssTccTimeout);
+		if(finalRangeTime <= stepTime) {
+			return false;
+		}
+
+		finalRangeTime -= stepTime;
+	
+	} // additional time for msrc is only needed if dss is disabled
+	else if(sequenceInfo.enabledSteps & MeasurementSequenceStep::MSRC)
+	{
+		const uint16_t stepTime = TimeOverhead::MSRC + msrcDssTccTimeout;
+		if(finalRangeTime <= stepTime) {
+			return false;
+		}
+
+		finalRangeTime -= stepTime;
+	}
+
+	if(sequenceInfo.enabledSteps & MeasurementSequenceStep::PreRange)
+	{
+		const uint16_t stepTime = TimeOverhead::PreRange
+			+ timeoutToMicroseconds(sequenceInfo.preRangeTimeout, sequenceInfo.vcselPeriodPreRange);
+
+		if(finalRangeTime <= stepTime) {
+			return false;
+		}
+
+		finalRangeTime -= stepTime;
+	}
+
+	if(finalRangeTime <= TimeOverhead::FinalRange) {
+		return false;
+	}
+
+	finalRangeTime -= TimeOverhead::FinalRange;
+
+	// convert value from microseconds to mclks
+	// from now on finalRangeTime is a value in mclks
+	finalRangeTime = timeoutToMclks(finalRangeTime, sequenceInfo.vcselPeriodFinalRange);
+
+	if(sequenceInfo.enabledSteps & MeasurementSequenceStep::PreRange) {
+		finalRangeTime += sequenceInfo.preRangeTimeout;
+	}
+
+	registerValue = encodeTimeout(finalRangeTime);
+	return true;
+}
+
+template < typename I2cMaster >
+uint32_t
+xpcc::Vl53l0<I2cMaster>::timeoutToMicroseconds(uint16_t mclks, uint8_t vcselPeriodPclks)
+{
+	// in nanoseconds
+	const uint32_t macroPeriodNs = ((UINT32_C(2304) * vcselPeriodPclks * 1655) + 500) / 1000;
```

Comment:
Why not use `uint32_t(2304)` here and below? I don't think you're outsmarting the compiler here ;-P